File: server/src/com/litbikes/model/PowerUp.java
package com.litbikes.model;

import java.awt.Rectangle;
import java.awt.geom.Line2D;
import java.util.UUID;

import com.litbikes.dto.PowerUpDto;
import com.litbikes.util.Vector;

public class PowerUp {
	public enum PowerUpType {
		NOTHING,
		ROCKET,
		SLOW
	}
	
	private String id;
	private Vector pos;
	private PowerUpType type;
	private boolean collected;
	
	public PowerUp(Vector _pos, PowerUpType _type) {
		id = UUID.randomUUID().toString();
		pos = _pos;
		type = _type;
		collected = false;
	}

	public String getId() {
		return id;
	}

	public Vector getPos() {
		return pos;
	}

	public void setPos(Vector pos) {
		this.pos = pos;
	}

	public PowerUpType getType() {
		return type;
	}

	public void setType(PowerUpType type) {
		this.type = type;
	}

	public boolean isCollected() {
		return collected;		
	}

	public void setCollected(boolean collected) {
		this.collected = collected;
	}
	
	public boolean collides(Line2D line) {
		int boxSize = 6;
		int originX = (int)pos.x - (boxSize / 2);
		int originY = (int)pos.y - (boxSize / 2);
		Rectangle hitbox = new Rectangle(originX, originY, boxSize, boxSize);
		boolean hits = line.intersects(hitbox);
		return line.intersects(hitbox);
	}
	
	public PowerUpDto getDto() {
		PowerUpDto dto = new PowerUpDto();
		dto.id = id;
		dto.pos = pos;
		dto.type = type;
		dto.collected = collected;
		return dto;
	}
}

File: client/src/model/powerUp.ts
module Model {
    import Vector = Util.Vector;
    import PowerUpDto = Dto.PowerUpDto;
    export class PowerUp {
        private id: string;
        private pos: Vector;
        private type: string;
        private collected: boolean;
        private collecting: boolean;
        private popSize: number;
        
        constructor(_id: string, _pos: Vector, _type: string) {
            this.id = _id;
            this.pos = _pos;
            this.type = _type;
            this.collected = false;
            this.collecting = false;
            this.popSize = 0;
        }
    
        public getId(): string {
            return this.id;
        }
            
        public getPos(): Vector {
            return this.pos;
        }
    
        public setPos(pos: Vector) {
            this.pos = pos;
        }
    
        public getType() {
            return this.type;
        }
    
        public setType(type: string) {
            this.type = type;
        }

        public isCollected(): boolean {
            return this.collected;
        }

        public collect() {
            this.collected = true;
            this.collecting = true;
        }

        public updateFromDto(dto: PowerUpDto) {
            this.pos = dto.pos;            
            if (!this.collected && dto.collected) {
                this.collect();
            }
        }

        public draw(p: p5) {
            if (!this.collected) {
                p.push();
                p.noStroke();
                p.translate(this.pos.x, this.pos.y);
                let size = 3;
                switch (this.type.toLowerCase()) {
                    case "rocket":
                        p.rotate(p.frameCount / -10.0);
                        p.fill('rgb(255,255,105)');
                        p.triangle(
                            -size, size * 0.8, 
                            size, size * 0.8, 
                            0, -size
                        );
                        break;
                    case "slow":
                        p.rotate(p.frameCount / 50.0);
                        p.fill('rgb(105,255,255)');
                        p.triangle(
                            -size, size, 
                            size, size, 
                            0, -size
                        );
                        p.rotate(p.PI);
                        p.triangle(
                            -size, size, 
                            size, size, 
                            0, -size
                        );
                        break;
                    default:
                        break;
                }
                p.pop();
            } else if (this.collecting) {
                p.fill('rgba(0,0,0,0)');
                p.stroke(255);
                p.strokeWeight(2);
                p.ellipse(this.pos.x, this.pos.y, this.popSize, this.popSize);
                this.popSize = this.popSize + 1.5;
                if (this.popSize > 20) {
                    this.collecting = false;
                }
            }
        }        
    }
}